sql/2024/04/image-formats.sql

```diff
@@ -0,0 +1,87 @@
+CREATE TEMPORARY FUNCTION
+  IS_CMS(technologies ARRAY<STRUCT<technology STRING,
+                                   categories ARRAY<STRING>,
+                                   info ARRAY<STRING>>>,
+         cms STRING,
+         version STRING)
+  RETURNS BOOL AS ( EXISTS(
+  SELECT
+    *
+  FROM
+    UNNEST(technologies) AS technology,
+    UNNEST(technology.info) AS info
+  WHERE
+      technology.technology = cms
+    AND ( version = ""
+    OR ENDS_WITH(version, ".x")
+            AND (STARTS_WITH(info, RTRIM(version, "x"))
+        OR info = RTRIM(version, ".x"))
+    OR info = version ) ) );
+
+WITH
+  pagesWithImages AS (
+    SELECT
+      date,
+      client,
+      page,
+      LOWER(JSON_EXTRACT_SCALAR(custom_metrics, '$.performance.lcp_elem_stats.nodeName')) = 'img' AS has_lcp_image,
+      LOWER(JSON_EXTRACT_SCALAR(custom_metrics, '$.performance.lcp_elem_stats.url')) = JSON_EXTRACT_SCALAR(images, '$.url') AS is_lcp_image,
+      JSON_EXTRACT_SCALAR(images, '$.url') AS url,
+      JSON_EXTRACT_SCALAR(images, '$.naturalWidth') AS image_width,
+      JSON_EXTRACT_SCALAR(images, '$.naturalHeight') AS image_height,
+      CAST(JSON_EXTRACT_SCALAR(images, '$.naturalWidth') AS NUMERIC) * CAST(JSON_EXTRACT_SCALAR(images, '$.naturalWidth') AS NUMERIC) AS image_dimensions
+    FROM
+      `httparchive.all.pages`,
+      UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.Images')) AS images
+    WHERE
+      IS_CMS(technologies,
+             'WordPress',
+             '')
+      AND date = '2024-02-01' ),
+
+  requests AS (
+    SELECT
+      date,
+      client,
+      page,
+      url,
+      LOWER(CAST(JSON_EXTRACT_SCALAR(summary, "$.mimeType") AS STRING)) AS mime_type,
+      CAST( JSON_EXTRACT_SCALAR(summary, "$.respSize") AS NUMERIC) AS resp_size,
+    FROM
+      `httparchive.all.requests`
+    WHERE
+        LOWER(CAST(JSON_EXTRACT_SCALAR(summary, "$.mimeType") AS STRING)) != ''
+      AND date = '2024-02-01' )
+
+SELECT
+  client,
+  is_lcp_image,
+  APPROX_QUANTILES(image_width, 1000)[
+    OFFSET
+      (500)] AS median_width,
+  APPROX_QUANTILES(image_height, 1000)[
+    OFFSET
+      (500)] AS median_height,
+  APPROX_QUANTILES(resp_size, 1000)[
+    OFFSET
+      (500)] AS median_file_size,
+  mime_type
+FROM
+  pagesWithImages
+    JOIN
+  requests
+  USING
+    ( page,
+      client,
+      url )
```

Comment:
While it's not really going to change the result because the dates are already limited to refer to the same hard-coded date, a good practice is to also include `date` here, as they must match. The very small real benefit even in this particular query is that this way you only get one `date` column in the combined result rather than `pagesWithImages.date` _and_ `requests.date`.